runtime-rs: Introduce PCIe Port devices in runtime-rs for qemu-rs

Apokleos opened the pull request:
To well support vfio devices in PCIe root ports or switch ports in case of qemu-rs, we instroduce port devices containing RootPort and SwitchPort mostly aligned with the implementations of kata-runtime.

BbolroC commented on src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs:
```diff
@@ -151,9 +152,51 @@ impl VirtSandbox {
+                port_configs.push(switch_port_config);
+            }
+            Some(port_configs)
+        } else {
```
Based on the following logic from `get_pcie_port()`:

```rust
pub fn get_pcie_port(&self) -> PCIePort {
        match (self.pcie_root_ports, self.pcie_switch_ports) {
            (0, 0) => PCIePort::NoPort,
            (r, s) if r > 0 && s > 0 => PCIePort::InvalidPort,
            (r, _) if r > 0 => PCIePort::RootPort,
            (_, s) if s > 0 => PCIePort::SwitchPort,
            _ => PCIePort::InvalidPort,
        }
    }
```
It seems a case for `r > 0 && s > 0` does not fall into this arm. I was wondering if I am missing something here.

Apokleos replied:
Here, it's used to prevent the case of both of root ports and switch ports set number which is greater than 0.
If it is removed, the (r, _) branch will contains case of (r>0, s>0), which is not allowed.
To clarify this function, test cases added to help understand.

BbolroC commented on src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs:
```diff
@@ -151,9 +152,51 @@ impl VirtSandbox {
 
+        // prepare pcie port device configs
+        if let Some(port_configs) = self.prepare_pcie_port_devices().await {
+            resource_configs.extend(port_configs);
```
Is there any reason to use `extend()` instead of `push()`?

Apokleos replied:
As `port_configs` is a list of port configs whose types are ResourceConfig, and `resource_configs` is also a list `vec![ResourceConfig]`, it will use extend instead of `push`.

BbolroC commented on src/runtime-rs/crates/hypervisor/src/device/topology.rs:
```diff
@@ -225,15 +304,35 @@ impl PCIeTopology {
+        let total_rp: usize = topo_config.device_info.pcie_root_port as usize;
+        let total_sp: usize = topo_config.device_info.pcie_switch_port as usize;
+
+        let mut devices_per_port: HashMap<PCIePort, Vec<TopologyPortDevice>> = HashMap::new();
```
Is it really necessary to always insert both `PCIePort` types even if we have the following logic below:

```rust
(r, s) if r > 0 && s > 0 => PCIePort::InvalidPort,
```

I am not sure, but it would be better to place a validity check before so that we could prepare either of them.

Apokleos replied:
validity check logic added

BbolroC commented on src/runtime-rs/crates/hypervisor/src/device/topology.rs:
```diff
@@ -301,6 +400,24 @@ impl PCIeTopology {
+    ) -> Option<TopologyPortDevice> {
+        if let Some(port_devices) = self.pcie_devices_per_port.get_mut(&port_type) {
+            for port in port_devices.iter_mut() {
+                info!(sl!(), "Checking port with device_id: {:?}.", &device_id);
```
`device_id` is only passed for logging?

BbolroC reviewed (commented):
Overall, it looks great to me! I've left a few minor questions.
Since I'm not an expert in this area, I would like to delegate the approval to @gkurz and @pmores. Thanks a lot!

zvonkok commented:
There are limits per HV how many root/switch ports can be created. We need to make sure that we're not exceeding those limits otherwise we get cryptic memory errors. 

For QEMU we have a limit of 16 root ports or 16 switch ports or 64KB IO memory each root or switch port taking up of default 4K IO memory.

Apokleos commented:
> There are limits per HV how many root/switch ports can be created. We need to make sure that we're not exceeding those limits otherwise we get cryptic memory errors.
> 
> For QEMU we have a limit of 16 root ports or 16 switch ports or 64KB IO memory each root or switch port taking up of default 4K IO memory.
>
Yes, sir. without KEP-4113 support, we need introduce many related codes to do such work which seems hacking to me.
But I am thinking how to address this problem and will make it be more friendly for the future KEP-4113 or sandbox API.

I have pulled request for get bar max memory in [another PR](https://github.com/kata-containers/kata-containers/pull/10579). And I am looking forward to your INPUT @zvonkok @BbolroC

zvonkok commented:
`hotplug_vfio_on_root_bus = true` this is a very old config option unsupported option. We removed that some time ago.  We should use 

```
hot_plug_vfio = "root-port"
cold_plug_vfo = "root-port"
```

zvonkok commented:
@Apokleos For the switch-port use-case I need to post a QEMU patch. Can you check if the BARs are correctly mapped? 
```
dmesg | grep BAR
```

Apokleos commented:
> > @BbolroC One outstanding thing to clarify is can we attach VFIO-AP devices to a root-port? This way we can drop the support for legacy bridges altogether.
> 
> Please take a look at the following:
> 
> * @hbrueckner 's comment: [Support basic capabilities to hotplug VFIO devices to root-port or switch-port for qemu-rs #10361 (comment)](https://github.com/kata-containers/kata-containers/issues/10361#issuecomment-2441468100)
> * [Add subchannel and VFIO-AP hot-plugging support to qemu-runtime-rs for s390x #10573](https://github.com/kata-containers/kata-containers/issues/10573)
> 
> Based on my understanding so far, there is no PCI support in s390x QEMU (so no concept like a `root/switch` port) and a bridge info is not required for VFIO-AP hot-plugging. I will raise a PR for VFIO-AP hot-plugging after this PR and #10362 are merged. Thanks.
>
Thx Choi @BbolroC, I'd like to know that there's some work in my PR I can do for you which can help your coming works ? If any ideas, please let me know.

Apokleos commented:
> `hotplug_vfio_on_root_bus = true` this is a very old unsupported option. We removed that some time ago. We should use
> 
> ```
> hot_plug_vfio = "root-port"
> cold_plug_vfio = "root-port"
> ```
>
Sure. the coming updated PR will remove this option.

zvonkok commented:
@Apokleos I am wondering how we're dealing with different archs, in the go-runtime we have different implementation for different architectures like x86 - s390x how are we dealing it with in runtime-rs? 
A good example is that in x86 we do not need the bridge but s390x needs it. So we shouldn't introduce the bridge in x86 but only have it enabled in s390x.

BbolroC commented:
> @Apokleos I am wondering how we're dealing with different archs, in the go-runtime we have different implementation for different architectures like x86 - s390x how are we dealing it with in runtime-rs? A good example is that in x86 we do not need the bridge but s390x needs it. So we shouldn't introduce the bridge in x86 but only have it enabled in s390x.

As I mentioned above, the bridge implementation for s390x in go-runtime is not necessary anymore in runtime-rs.  https://github.com/kata-containers/kata-containers/pull/10574 for runtime-rs covers all for s390x and the bridge is not needed for hot-plugging VFIO-AP.

pmores commented on src/runtime-rs/crates/hypervisor/src/device/topology.rs:
```diff
@@ -225,15 +304,38 @@ impl PCIeTopology {
         })
     }
 
+    pub fn get_pcie_port(&self) -> PCIePort {
```
This function looks useful but I don't see it called from anywhere.  If that's the case actually, do we need it?

pmores commented on src/runtime-rs/config/configuration-qemu-runtime-rs.toml.in:
```diff
@@ -392,6 +392,13 @@ pflashes = []
+# The value means how many devices attached onto pcie_switch_port will be created.
+# This value is valid when hotplug_vfio_on_root_bus is true, and machine_type is "q35"
+# Default 0
+#pcie_switch_port = 2
```
Seeing as it's illegal to have to both `pcie_root_port` and `pcie_switch_port` set to non-zero values, shouldn't we document this limitation for our users?

pmores replied:
I based my comment on the final hunk of the final commit of this series which seems to cause the config file validation to fail if both are non-zero.  Unfortunately I know too little about actual PCIe topologies to decide whether that's fair or not. :-)

pmores replied:
I've approved already but I'd like to clarify - the check making it illegal to have both `pcie_root_port` and `pci_switch_port` non-zero was removed from code so this suggestion of mine no longer applies?

pmores replied:
Ah okay, so the check is still there though in a different commit.  I'd prefer this limitation to be documented in `configuration-qemu-runtime-rs.toml.in` for the sake of our users, to avoid confusing them.

pmores commented on src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs:
```diff
@@ -1700,6 +1701,182 @@ impl ToQemuParams for ObjectIoThread {
+impl ToQemuParams for PCIeRootPortDevice {
+    async fn qemu_params(&self) -> Result<Vec<String>> {
+        let mut qemu_params = Vec::new();
+        let mut device_params = Vec::new();
```
Is there any advantage to using two vectors over the idiom used elsewhere in the file which allocates a single vector at the top of the function
```
let mut params = Vec::new();
```
and finishes after filling it in with
```
Ok(vec!["-device".to_owned(), params.join(",")])
```
?

gkurz commented on src/runtime-rs/crates/hypervisor/src/dragonball/inner_device.rs:
```diff
@@ -96,7 +96,7 @@ impl DragonballInner {
                 Ok(DeviceType::VhostUserNetwork(dev))
             }
-            DeviceType::Vsock(_) => todo!(),
+	    DeviceType::Vsock(_) | DeviceType::PortDevice(_) => todo!(),
```
```
standard rust check...
cargo fmt -- --check
Diff in /run/media/greg/GrougSSD/Work/kata-containers/kata-containers/src/runtime-rs/crates/hypervisor/src/dragonball/inner_device.rs at line 96:
                     .context("add vhost-user-net device")?;
                 Ok(DeviceType::VhostUserNetwork(dev))
             }
-	    DeviceType::Vsock(_) | DeviceType::PortDevice(_) => todo!(),
+            DeviceType::Vsock(_) | DeviceType::PortDevice(_) => todo!(),
         }
     }
 ```

Please convert the tab character to spaces.

gkurz reviewed (commented):
Hi Alex ! Thanks for working on this ambitious addition.

This first round of review is, as usual, focused on verifying that individual commits don't have obvious flaws.

I found some I'd like to see fixed, especially the ordering problem that breaks bisect.
Also it would be appreciated to have descriptions with informative non-obvious details in all commits. For example, the justification to not allow root ports and switch ports at the same time (why?) should be documented in some commit description.

pmores commented on src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs:
```diff
@@ -1700,6 +1701,182 @@ impl ToQemuParams for ObjectIoThread {
+    ) -> Self {
+        PCIeRootPortDevice {
+            id: id.to_string(),
+            bus: bus.to_string(),
```
To comply with the conventional constructor postcondition and guarantee that on constructor exit, the instance is fully initialised, I think we'd better do something like
```
bus: if bus.is_empty() { DEFAULT_PCIE_ROOT_BUS.to_owned() } else { bus.to_string() },
```
here.  The same goes for the rest of the fields that represent mandatory parameters (i.e. all but `mem_reserve` and `pref64_reserve`).  This would also make `qemu_params()` implementation more straightfoward, just serialising members to the command line unconditionally.

pmores commented on src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs:
```diff
@@ -1700,6 +1701,182 @@ impl ToQemuParams for ObjectIoThread {
+            multifunction: multifunc,
+            addr: addr.to_string(),
+            mem_reserve: format!("{}B", mem_reserve),
+            pref64_reserve: format!("{}B", pref64_reserve),
```
`qemu_params()` implementation suggests that the two `*_reserve` params are not mandatory.  If that's indeed the case (I'm not sure since while `qemu_params()` serialises them only conditionally, if they aren't empty, this constructor implementation seems to ensure that they are actually never empty), I think we'd better drop them from the constructor params list and provide setters instead, e.g.:
```
    fn set_mem_reserve(&mut self, mem_reserve: u64) -> &mut Self {
        self.mem_reserve = mem_reserve;
        self
    }
```
The call site then becomes something like
```
let mut root_port_dev = PCIeRootPortDevice::new(...);
root_port_dev
    .set_mem_reserve(...)
    .set_pref64_reserve(...);
```
I understand this makes it seem a bit more verbose but I think the added semantic clarity is worth it.

Also, dropping the two params from the `new()` params list would apparently let us leave the `clippy::too_many_arguments` warning on. :-)

pmores commented on src/runtime-rs/crates/hypervisor/src/device/topology.rs:
```diff
@@ -301,6 +403,20 @@ impl PCIeTopology {
         }
     }
 
+    pub fn find_vacant_port(&mut self, port_type: PCIePort) -> Option<TopologyPortDevice> {
```
This function doesn't seem to be called either... are we missing a call to it from elsewhere in the PR?

pmores commented:
Hey @Apokleos , thanks for the effort you've put into this PR, it's very well structured and despite the non-trivial subject matter extremely easy to read and review!

I've left here a bunch of comments and questions but I can see no fundamental problems and overall I'm very keen to get this merged.

One general question I'd like to ask is, the topology changes seem strangely underused.  Apart from the functions seemingly not called which I pointed out near the corresponding hunks, also `PCIeTopology::pcie_devices_per_port` seems to be filled in in `PCIePortDevice::attach()` but apparently not touched afterwards.  If you could clarify the purpose of those changes I'd appreciate that.

kevinzs2048 commented:
Any progress on this PR recently? On arm64/QemuVirt we are relying on this PR to compose the right PCIE topology in order to get the sandbox work.

Apokleos commented:
@zvonkok @pmores @gkurz @kevinzs2048 PTAL !

Apokleos commented:
- Switch ports:
```bash
/opt/kata/bin/qemu-system-x86_64 -name sandbox-00f56906779642cb8e3aa45a46d1ad8d9558b6b32fb0308bb450e7e72e9c9241 -kernel /opt/kata/share/kata-containers/vmlinux-6.12.13-150 -append reboot=k panic=1 systemd.unit=kata-containers.target systemd.mask=systemd-networkd.service root=/dev/pmem0p1 rootflags=dax,data=ordered,errors=remount-ro ro rootfstype=ext4 cgroup_no_v1=all systemd.unified_cgroup_hierarchy=1 selinux=0 -smp 1,maxcpus=128 -machine q35,accel=kvm,nvdimm=on -cpu host,pmu=off -m 2G,slots=128,maxmem=515511M -object memory-backend-file,id=entire-guest-memory-share,mem-path=/dev/shm,size=2G,share=on,readonly=off -qmp unix:fd=16,server=on,wait=off -rtc base=utc,clock=host,driftfix=slew -object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0 -device pci-bridge,bus=pcie.0,id=pci-bridge-0,chassis_nr=1,shpc=off,addr=2,io-reserve=4k,mem-reserve=1m,pref64-reserve=1m -device pci-bridge,bus=pcie.0,id=pci-bridge-1,chassis_nr=2,shpc=off,addr=3,io-reserve=4k,mem-reserve=1m,pref64-reserve=1m -device virtio-scsi-pci,id=scsi0 -device vhost-vsock-pci,vhostfd=17,guest-cid=4273688281 -chardev socket,id=virtiofsd-chardev,path=virtiofsd.sock -device vhost-user-fs-pci,chardev=virtiofsd-chardev,tag=kataShared -numa node,memdev=entire-guest-memory-share -object memory-backend-file,id=TODO,mem-path=/opt/kata/share/kata-containers/kata-ubuntu-jammy.image,size=256M,share=off,readonly=on -device nvdimm,memdev=TODO,unarmed=on -device pcie-root-port,id=rp1,bus=pcie.0,chassis=2,slot=1,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device x3130-upstream,id=swup1,bus=rp1 -device xio3130-downstream,id=swdp5,bus=swup1,chassis=8,slot=5 -device xio3130-downstream,id=swdp6,bus=swup1,chassis=9,slot=6 -device xio3130-downstream,id=swdp3,bus=swup1,chassis=6,slot=3 -device xio3130-downstream,id=swdp1,bus=swup1,chassis=4,slot=1 -device xio3130-downstream,id=swdp8,bus=swup1,chassis=11,slot=8 -device xio3130-downstream,id=swdp2,bus=swup1,chassis=5,slot=2 -device xio3130-downstream,id=swdp4,bus=swup1,chassis=7,slot=4 -device xio3130-downstream,id=swdp7,bus=swup1,chassis=10,slot=7 -device virtio-serial-pci,id=serial0 -device virtconsole,id=console0,chardev=charconsole0 -chardev socket,id=charconsole0,server=on,wait=off,path=console.sock -vga none -no-user-config -nodefaults -nographic -no-reboot

$sudo nerdctl run --runtime io.containerd.kata.v2 -it --rm ubuntu-lspci:22.04
root@00f569067796:/# lspci -t
-[0000:00]-+-00.0
           +-01.0
           +-02.0-[01]----00.0
           +-03.0-[02]--
           +-04.0
           +-05.0
           +-06.0
           +-07.0-[03-0c]----00.0-[04-0c]--+-00.0-[05]--
           |                               +-01.0-[06]--
           |                               +-02.0-[07]--
           |                               +-03.0-[08]--
           |                               +-04.0-[09]--
           |                               +-05.0-[0a]--
           |                               +-06.0-[0b]--
           |                               \-07.0-[0c]--
           +-08.0
           +-1f.0
           +-1f.2
           \-1f.3
root@00f569067796:/# lspci
00:00.0 Host bridge: Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
00:01.0 Unclassified device [00ff]: Red Hat, Inc. Virtio RNG
00:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:03.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:04.0 SCSI storage controller: Red Hat, Inc. Virtio SCSI
00:05.0 Communication controller: Red Hat, Inc. Virtio socket (rev 01)
00:06.0 Mass storage controller: Red Hat, Inc. Virtio file system (rev 01)
00:07.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:08.0 Communication controller: Red Hat, Inc. Virtio console
00:1f.0 ISA bridge: Intel Corporation 82801IB (ICH9) LPC Interface Controller (rev 02)
00:1f.2 SATA controller: Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode] (rev 02)
00:1f.3 SMBus: Intel Corporation 82801I (ICH9 Family) SMBus Controller (rev 02)
01:00.0 Ethernet controller: Red Hat, Inc. Virtio network device
03:00.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Upstream) (rev 02)
04:00.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
04:01.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
04:02.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
04:03.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
04:04.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
04:05.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
04:06.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
04:07.0 PCI bridge: Texas Instruments XIO3130 PCI Express Switch (Downstream) (rev 01)
```

- Root Ports:
```bash

/opt/kata/bin/qemu-system-x86_64 -name sandbox-f9bcc769abcc2b9295809418be456057a00cb7559780470d121d86f66345dee9 -kernel /opt/kata/share/kata-containers/vmlinux-6.12.13-150 -append reboot=k panic=1 systemd.unit=kata-containers.target systemd.mask=systemd-networkd.service root=/dev/pmem0p1 rootflags=dax,data=ordered,errors=remount-ro ro rootfstype=ext4 cgroup_no_v1=all systemd.unified_cgroup_hierarchy=1 selinux=0 -smp 1,maxcpus=128 -machine q35,accel=kvm,nvdimm=on -cpu host,pmu=off -m 2G,slots=128,maxmem=515511M -object memory-backend-file,id=entire-guest-memory-share,mem-path=/dev/shm,size=2G,share=on,readonly=off -qmp unix:fd=16,server=on,wait=off -rtc base=utc,clock=host,driftfix=slew -object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0 -device pci-bridge,bus=pcie.0,id=pci-bridge-0,chassis_nr=1,shpc=off,addr=2,io-reserve=4k,mem-reserve=1m,pref64-reserve=1m -device pci-bridge,bus=pcie.0,id=pci-bridge-1,chassis_nr=2,shpc=off,addr=3,io-reserve=4k,mem-reserve=1m,pref64-reserve=1m -device virtio-scsi-pci,id=scsi0 -device vhost-vsock-pci,vhostfd=17,guest-cid=2690525519 -chardev socket,id=virtiofsd-chardev,path=virtiofsd.sock -device vhost-user-fs-pci,chardev=virtiofsd-chardev,tag=kataShared -numa node,memdev=entire-guest-memory-share -object memory-backend-file,id=TODO,mem-path=/opt/kata/share/kata-containers/kata-ubuntu-jammy.image,size=256M,share=off,readonly=on -device nvdimm,memdev=TODO,unarmed=on -device pcie-root-port,id=rp4,bus=pcie.0,chassis=5,slot=4,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp5,bus=pcie.0,chassis=6,slot=5,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp6,bus=pcie.0,chassis=7,slot=6,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp1,bus=pcie.0,chassis=2,slot=1,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp0,bus=pcie.0,chassis=1,slot=0,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp2,bus=pcie.0,chassis=3,slot=2,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp7,bus=pcie.0,chassis=8,slot=7,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp3,bus=pcie.0,chassis=4,slot=3,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device virtio-serial-pci,id=serial0 -device virtconsole,id=console0,chardev=charconsole0 -chardev socket,id=charconsole0,server=on,wait=off,path=console.sock -vga none -no-user-config -nodefaults -nographic -no-reboot

$sudo nerdctl run --runtime io.containerd.kata.v2 -it --rm ubuntu-lspci:22.04
root@f9bcc769abcc:/# lspci
00:00.0 Host bridge: Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
00:01.0 Unclassified device [00ff]: Red Hat, Inc. Virtio RNG
00:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:03.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:04.0 SCSI storage controller: Red Hat, Inc. Virtio SCSI
00:05.0 Communication controller: Red Hat, Inc. Virtio socket (rev 01)
00:06.0 Mass storage controller: Red Hat, Inc. Virtio file system (rev 01)
00:07.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:08.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:09.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:0a.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:0b.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:0c.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:0d.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:0e.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:0f.0 Communication controller: Red Hat, Inc. Virtio console
00:1f.0 ISA bridge: Intel Corporation 82801IB (ICH9) LPC Interface Controller (rev 02)
00:1f.2 SATA controller: Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode] (rev 02)
00:1f.3 SMBus: Intel Corporation 82801I (ICH9 Family) SMBus Controller (rev 02)
01:00.0 Ethernet controller: Red Hat, Inc. Virtio network device
root@f9bcc769abcc:/# lspci -t
-[0000:00]-+-00.0
           +-01.0
           +-02.0-[01]----00.0
           +-03.0-[02]--
           +-04.0
           +-05.0
           +-06.0
           +-07.0-[03]--
           +-08.0-[04]--
           +-09.0-[05]--
           +-0a.0-[06]--
           +-0b.0-[07]--
           +-0c.0-[08]--
           +-0d.0-[09]--
           +-0e.0-[0a]--
           +-0f.0
           +-1f.0
           +-1f.2
           \-1f.3
```

pmores commented on src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs:
```diff
@@ -1824,6 +1825,178 @@ impl ToQemuParams for ObjectSevSnpGuest {
+            DEFAULT_PCIE_ROOT_BUS
+        } else {
+            &self.bus
+        };
```
We have done this in `new()` already, I don't think `self.bus` can be empty by now.

Apokleos replied:
Yes, you're right. fixed.

pmores commented on src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs:
```diff
@@ -1824,6 +1825,178 @@ impl ToQemuParams for ObjectSevSnpGuest {
+            device_params.push(format!("addr={}", addr));
+        }
+
+        device_params.push(format!("multifunction={}", multifunction));
```
Frankly I'd prefer for `chassis`, `slot`, `multifunction` and `addr` to be treated similar to `bus`, i.e. do the checking and default value substitution as early as possible, in `new()` since all of them are non-optional params of `new()`.

However I don't see this as blocking a merge, I can make these adjustments later on if we agree on doing so.

Apokleos replied:
> Frankly I'd prefer for `chassis`, `slot`, `multifunction` and `addr` to be treated similar to `bus`, i.e. do the checking and default value substitution as early as possible, in `new()` since all of them are non-optional params of `new()`.
> 
Good cache ! I think this is a very constructive and helpful suggestion. fixed

pmores replied:
Thanks!

pmores commented on src/runtime-rs/crates/hypervisor/src/device/topology.rs:
```diff
@@ -203,14 +203,143 @@ pub struct PCIeRootComplex {
+        }
+    }
+
+    pub fn bus(&self) -> String {
```
Maybe it's just me but I find it a bit confusing to have both `TopologyPortDevice::bus` and `TopologyPortDevice::bus()`, meaning different things.  Could one of them be better called something else?

Apokleos replied:
Sorry here, It's indeed misleading and I correct it with `port_id`, does it make sense ?

pmores replied:
I intentionally refrained from making a specific suggestion since I'm not familiar enought with PCIe terminology but `port_id` makes a lot of sense to me.  Thanks!

pmores commented on src/runtime-rs/crates/hypervisor/src/device/topology.rs:
```diff
@@ -203,14 +203,143 @@ pub struct PCIeRootComplex {
+    SwitchPort,
+
+    // InvalidPort is for invalid port
+    InvalidPort,
```
I don't see any usage of `InvalidPort`, perhaps it could be removed?

Apokleos replied:
fixed

pmores replied:
Thanks!

kevinzs2048 commented:
@Apokleos Thanks a lot Alex. I also rebased my path on top of your newest PRs.
Here are some test results on Arm64 platform, I'm using the VM on Apple M4 for development(nested virtualization):
```
/home/kevin/qemu-9.0.4/build/qemu-system-aarch64 -name sandbox-338092812a540dfc21e5a6b79f98c0fe39afeffc8c6a1d973100834d92ad2e79 -kernel /usr/share/kata-containers/vmlinux-6.12.13-149 -append reboot=k panic=1 systemd.unit=kata-containers.target systemd.mask=systemd-networkd.service agent.log_vport=1025 root=/dev/vda1 rootflags=data=ordered,errors=remount-ro ro rootfstype=ext4 agent.log=debug agent.debug_console agent.debug_console_vport=1026 agent.log=debug initcall_debug selinux=0 -smp 1,maxcpus=4 -machine virt,accel=kvm,nvdimm=on,usb=off,gic-version=3 -cpu host,pmu=off -m 2G,slots=128,maxmem=15957M -object memory-backend-file,id=entire-guest-memory-share,mem-path=/dev/shm,size=2G,share=on,readonly=off -qmp unix:fd=16,server=on,wait=off -rtc base=utc,clock=host,driftfix=slew -object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0 -device pci-bridge,bus=pcie.0,id=pci-bridge-0,chassis_nr=1,shpc=off,addr=2,io-reserve=4k,mem-reserve=1m,pref64-reserve=1m -device vhost-vsock-pci,vhostfd=17,guest-cid=27053356 -chardev socket,id=virtiofsd-chardev,path=virtiofsd.sock -device vhost-user-fs-pci,chardev=virtiofsd-chardev,tag=kataShared -numa node,memdev=entire-guest-memory-share -blockdev driver=file,node-name=image-4051f611e7e5e93f,filename=/home/kevin/opt/kata/share/kata-containers/kata-ubuntu-jammy.image,aio=threads,cache.direct=off,cache.no-flush=off,auto-read-only=on -device virtio-blk-pci,drive=image-4051f611e7e5e93f,config-wce=off,share-rw=on,serial=image-4051f611e7e5e93f -device pcie-root-port,id=rp1,bus=pcie.0,chassis=2,slot=1,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device pcie-root-port,id=rp0,bus=pcie.0,chassis=1,slot=0,multifunction=off,mem-reserve=33554432B,pref64-reserve=536870912B -device virtio-serial-pci,id=serial0 -device virtconsole,id=console0,chardev=charconsole0 -chardev socket,id=charconsole0,server=on,wait=off,path=console.sock -bios /home/kevin/qemu-9.0.4/build/pc-bios/edk2-aarch64-code.fd -vga none -no-user-config -nodefaults -nographic -no-reboot
```

```
sudo nerdctl run  --runtime io.containerd.kata.v2 -it --rm  ubuntu22-pci:latest  /bin/bash
root@338092812a54:/# lspci
00:00.0 Host bridge: Red Hat, Inc. QEMU PCIe Host bridge
00:01.0 Unclassified device [00ff]: Red Hat, Inc. Virtio RNG
00:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
00:03.0 Communication controller: Red Hat, Inc. Virtio socket (rev 01)
00:04.0 Mass storage controller: Red Hat, Inc. Virtio file system (rev 01)
00:05.0 SCSI storage controller: Red Hat, Inc. Virtio block device
00:06.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:07.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:08.0 Communication controller: Red Hat, Inc. Virtio console
03:00.0 Ethernet controller: Red Hat, Inc. Virtio network device (rev 01)

root@338092812a54:/# lspci -t
-[0000:00]-+-00.0
           +-01.0
           +-02.0-[01]--
           +-03.0
           +-04.0
           +-05.0
           +-06.0-[02]--
           +-07.0-[03]----00.0
           \-08.0
```

zvonkok commented:
Let's merge this first and not overload it with my work on NUMA. This is the base for implementing the VRA for GPUDirect RDMA and GPUDirect GDS.

zvonkok reviewed (approved):
LGTM, as a first iteration towards properly enabling GPUDirect RDMA, GDS. amazing work @Apokleos

BbolroC commented on src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs:
```diff
@@ -2192,6 +2354,137 @@ impl<'a> QemuCmdLine<'a> {
+        let machine_type: &str = &self.config.machine_info.machine_type;
+        let (addr, multi_function) = match machine_type {
+            "q35" | "virt" => ("0", false),
+            _ => return Ok(()),
```
Why don't we have log message here:

```rust
            _ => {
                info!(sl!(), "PCIe root ports not supported for machine type: {}", machine_type);
                return Ok(());
            },
```

BbolroC commented on src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs:
```diff
@@ -2192,6 +2354,137 @@ impl<'a> QemuCmdLine<'a> {
+
+        let machine_type = &self.config.machine_info.machine_type;
+        if machine_type != "q35" && machine_type != "virt" {
+            return Ok(());
```
same here

```rust
            info!(sl!(), "PCIe switch ports not supported for machine type: {}", machine_type);
            return Ok(());
```

Apokleos replied:
fixed

BbolroC reviewed (approved):
Thanks for reflecting my 1st review. I've left some comments regarding log messages, but they are something nice to have. So you can ignore them. I appreciate your ground work, @Apokleos.

pmores commented on src/runtime-rs/crates/hypervisor/src/device/topology.rs:
```diff
@@ -203,14 +203,139 @@ pub struct PCIeRootComplex {
+
+/// Represents an available node in the PCIe topology.
+#[derive(Clone, Debug)]
+pub enum AvailabledNode {
```
Should this be `AvailableNode`?

pmores reviewed (approved):
Thanks @Apokleos for this huge piece of work in a well-formatted PR! :-)  I'm approving but frankly I'd prefer if you could deal with some of the remaining suggestions by @gkurz and @BbolroC either by incorporating them and resolving the pertinent conversations, or by documenting your reasons why not to incorporate them.

I've resolved my conversations for clarity that you've dealt with.

Apokleos commented:
> Thanks @Apokleos for this huge piece of work in a well-formatted PR! :-) I'm approving but frankly I'd prefer if you could deal with some of the remaining suggestions by @gkurz and @BbolroC either by incorporating them and resolving the pertinent conversations, or by documenting your reasons why not to incorporate them.
> 
> I've resolved my conversations for clarity that you've dealt with.
>
Thx Pavel @pmores. 
Yes, absolutely I will do.  As I have promissed @zvonkok I will clean up these left comments.

Apokleos commented:
> `hotplug_vfio_on_root_bus = true` this is a very old unsupported option. We removed that some time ago. We should use
> 
> ```
> hot_plug_vfio = "root-port"
> cold_plug_vfio = "root-port"
> ```
>
I'd like to create an issue to do this work


> `hotplug_vfio_on_root_bus = true` this is a very old unsupported option. We removed that some time ago. We should use
> 
> ```
> hot_plug_vfio = "root-port"
> cold_plug_vfio = "root-port"
> ```
>
I think we should address it in another PR, ahead of it, I open [an issue](https://github.com/kata-containers/kata-containers/issues/11198) to track it

Apokleos commented on src/runtime-rs/crates/hypervisor/src/device/driver/port_device.rs:
```diff
@@ -0,0 +1,128 @@
+
+//use std::collections::HashMap;
+
+use anyhow::{anyhow, Context, Result};
```
Thx @gkurz  I have correct errors related to commit orders as your suggestion.

Apokleos commented:
Hi @zvonkok @pmores @BbolroC I have updated the PR, and try to address the left comments. Is there any stuff I need address before it's merged ?  It is with `do-not-merge` before your ACK.

BbolroC commented:
Thanks for the update. I am fine with merging this. 👍

pmores commented:
Hey @Apokleos , I'd still be in favour of mentioning the root/switch port counts limitation in the config file template but I'm fine merging this with or without that change.  Thanks!

zvonkok commented:
@Apokleos What happens if a VFIO device is hot-plugged to a root-port and the workload fails to start and the container is restarted. Is the root-port freed up or can it be reused once he container restarts?

pmores commented:
@zvonkok I don't think this PR deals with hotplugging at all.

Apokleos commented:
> Hey @Apokleos , I'd still be in favour of mentioning the root/switch port counts limitation in the config file template but I'm fine merging this with or without that change. Thanks!
>
Thx Pavel @pmores 
I hope my explanation makes sense. When I initially designed this, the primary goal was to align with the runtime-go's implementation to ensure functional consistency. 

Secondly, it's aimed for a simpler pcie topology that could still support advanced features for devices like GPUs. It's what the limitation only one port device type(RootPort or SwitchPort<combine of RootPort and SwitchPort>) exists in a pod/VM.

I also spent some time considering if there were scenarios requiring both a dedicated `root port` and a combined root/switch port or called `switch port`, but couldn't identify any clear use cases. That's what led to the current design. 

However, I'm certainly open to making adjustments in future implementations if such scenarios do arise.

Apokleos commented:
> @Apokleos What happens if a VFIO device is hot-plugged to a root-port and the workload fails to start and the container is restarted. Is the root-port freed up? Can it be reused once he container restarts?
>
Thx @zvonkok This PR will not involve the vfio device hotplugging stuff, and[ the following one](https://github.com/kata-containers/kata-containers/pull/10362) will do. But IMO, it's a point worth discussing. As I understand, continuously hot-plugging devices from VM isn't very reliable at the moment, so we might need to have a deep discussion about it based on the PR #10362.
